openedx/core/djangoapps/user_api/views.py

```diff
@@ -345,10 +350,63 @@ def post(self, request):
             }
             return JsonResponse(errors, status=400)
 
+        password = data.get("password")
+        password_copy = data.get("password_copy")
+        if password and password_copy and password != password_copy:
+            errors = {
+                'password_copy': [{"user_message": "Passwords don't match."}]
+            }
+            return JsonResponse(errors, status=400)
+
         response = JsonResponse({"success": True})
         set_logged_in_cookies(request, response, user)
+
+        if getattr(settings, 'REGISTRATION_EMAIL_FULL_VERIFICATION', None) is not None:
+            email = data.get("email")
+            error = self._verify_email_really_exists(email)
+            if error:
+                errors = {
+                    'email': [{"user_message": error}]
+                }
+                return JsonResponse(errors, status=400)
+
         return response
 
+    def _verify_email_really_exists(self, email):
+        """Check if a email really exists,
+        returns None if success, otherwise an error message
+        """
+        mx_record = None
+        try:
+            domain = email.rsplit('@', 1)[-1]
```

Comment:
@schenedx @roderick, will the following pull request be acceptable?
In the sign-in form change "Email" to "Email or username". If logging-in with email fails then try to find a user by username (username = params["email"]). If such a user is found then get the user's email and repeat logging-in with this email and params["password"]. Currently LMS doesn't allow to create users with the same username so usernames should be unique as emails.